Add VM architecture diagram (#8870)
istio/istio.io: content/en/docs/ops/deployment/vm-architecture/index.md

---
title: Virtual Machine Architecture
description: Describes Istio's high-level architecture for virtual machines.

> weight: 20

Reviewer: ```suggestion
weight: 25
```

Maybe its 15 so it's placed between Architecture and Deployment Models (already has a 20). Or maybe since you want them to be read first, make it 25. The current placement in the netlify shows it after Deployment models.

Reviewer 2: I still  see them after deployment models in rendered docs.

keywords:
- virtual-machine
test: n/a
owner: istio/wg-environments-maintainers
---

Before reading this document, be sure to review [Istio's architecture](/docs/ops/deployment/architecture/) and [deployment models](/docs/ops/deployment/deployment-models/).
This page builds on these pages to explain how Istio can be extended to support joining virtual machines into the mesh.

Istio's virtual machine expansion allows connecting workloads outside of a Kubernetes cluster to the mesh.

> This enables legacy applications, or applications not suitable to run in a containerized environment, to get all the same benefits that Istio provides to applications running inside Kubernetes.

Reviewer: ```suggestion
This enables legacy applications, or applications not suitable to run in a containerized environment, to get all the benefits that Istio provides to applications running inside Kubernetes.
```

You could remove all and leave same.

> Many of the features the Kubernetes automatically provides applications are supplemented by Istio.

Reviewer 2: Here's my suggested rewording:

For workloads running on Kubernetes, the Kubernetes platform itself provides various functionalities like service discovery, DNS resolution, and health checks which are often missing in virtual machine environments. Istio enables these features for workloads running on virtual machines, and in addition allows these workloads to utillize Istio functionalities like mutual TLS (mTLS), rich telemetry and, advanced traffic management capabilities.

{{< image width="75%"
    link="single-network.svg"
    alt="A service mesh with a single network and virtual machines"
    title="Single network"
    caption="A service mesh with a single network and virtual machines"
    >}}

{{< /tab >}}

{{< tab name="Multi-Network" category-value="multiple" >}}

In this mesh, there are multiple [networks](/docs/ops/deployment/deployment-models/#network-models), implying that pods and virtual machines are not able to communicate directly with each other.

Control plane traffic, including XDS configuration and certificate signing, are sent through a Gateway in the cluster.
Similarly, all communication between pods and virtual machines goes through the gateway, which acts as a bridge between the two networks.

{{< image width="75%"
    link="multi-network.svg"
    alt="A service mesh with multiple networks and virtual machines"
    title="Multiple networks"
    caption="A service mesh with multiple networks and virtual machines"
    >}}

{{< /tab >}}

{{< /tabset >}}

## Service association

Istio provides two mechanisms to represent virtual machine workloads:

> * [`WorkloadGroup`](/docs/reference/config/networking/workload-group/) represent a logical group of virtual machine workloads that share common properties. This is similar to a `Deployment` in Kubernetes.

Reviewer: ```suggestion
* [`WorkloadGroup`](/docs/reference/config/networking/workload-group/) represents a logical group of virtual machine workloads that share common properties. This is similar to a `Deployment` in Kubernetes.
```

> * [`WorkloadEntry`](/docs/reference/config/networking/workload-entry/) represent a single instance of a virtual machine workload. This is similar to a `Pod` in Kubernetes.

Reviewer: ```suggestion
* [`WorkloadEntry`](/docs/reference/config/networking/workload-entry/) represents a single instance of a virtual machine workload. This is similar to a `Pod` in Kubernetes.
```

kind: WorkloadEntry
metadata:
  annotations:
    istio.io/autoRegistrationGroup: product-vm
  labels:
    app: product
  name: product-vm-1.2.3.4
spec:
  address: 1.2.3.4
  labels:
    app: product
  serviceAccount: default
{{< /text >}}

This `WorkloadEntry` resource describes a single instance of a workload, similar to a pod in Kubernetes. When the workload is removed from the mesh, the `WorkloadEntry` resource will
be automatically removed.  Additionally, if any probes are configured in the `WorkloadGroup` resource, the Istio control plane automatically updates the health status of associated `WorkloadEntry` instances.

In order for consumers to reliably call your workload, it's recommended to declare a `Service` association. This allows clients to reach a stable hostname, like `product.default.svc.cluster.local`, rather than ephemeral IP addresses. This also enables you to use advanced routing capabilities in Istio via `DestinationRule` and `VirtualService` APIs.

Any Kubernetes service can transparently select workloads across pods and virtual machines via the selector fields which are matched with pod and `WorkloadEntry` labels respectively.

> For example, for a `Service` named `product` that was composed of a `Pod` and a `WorkloadEntry`:

Reviewer: ```suggestion
For example, a `Service` named `product` is composed of a `Pod` and a `WorkloadEntry`:
```

{{< image width="30%"
    link="service-selector.svg"
    title="Service Selection"
    >}}

> With this configuration, requests to `product` would be sent to both the pod and virtual machine workload.

Reviewer 2: ```suggestion
With this configuration, requests to `product` would be load-balanced to both the pod and virtual machine workload instances.
```

## DNS

> Kubernetes provides DNS resolution in pods for `Service` names  allowing pods to easily communicate with one another by stable hostnames.

Reviewer: ```suggestion
Kubernetes provides DNS resolution in pods for `Service` names allowing pods to easily communicate with one another by stable hostnames.
```

Istio's virtual machine expansion allows connecting workloads outside of a Kubernetes cluster to the mesh.
This enables legacy applications, or applications not suitable to run in a containerized environment, to get all the same benefits that Istio provides to applications running inside Kubernetes.

Many of the features the Kubernetes automatically provides applications are supplemented by Istio.
This includes service discovery, DNS resolution, and health checks.
Additionally, standard Istio features such as automatic mutual TLS, rich telemetry, and expressive traffic management configuration are enabled.

The following diagram shows the architecture of a mesh with virtual machines:

{{< tabset category-name="network-mode" >}}

{{< tab name="Single-Network" category-value="single" >}}

In this mesh, there is a single [network](/docs/ops/deployment/deployment-models/#network-models), implying that pods and virtual machines can communicate directly with each other.

Control plane traffic, including XDS configuration and certificate signing, are sent through a Gateway in the cluster.

> This ensures that the virtual machines have a stable address to connect to when they are bootstrapping. All other communication between pods and
> virtual machines are able to communicate directly.

Author: ```suggestion
This ensures that the virtual machines have a stable address to connect to when they are bootstrapping. Pods and virtual machines can communicate directly with each other without requiring any intermediate Gateway.
```

---
title: Virtual Machine Architecture
description: Describes Istio's high-level architecture for virtual machines.
weight: 20
keywords:
- virtual-machine
test: n/a
owner: istio/wg-environments-maintainers
---

Before reading this document, be sure to review [Istio's architecture](/docs/ops/deployment/architecture/) and [deployment models](/docs/ops/deployment/deployment-models/).
This page builds on these pages to explain how Istio can be extended to support joining virtual machines into the mesh.

Istio's virtual machine support involves connecting workloads outside of a Kubernetes cluster to the mesh.
This enables legacy applications, or applications not suitable to run inside of Kubernetes, to get all the same benefits Istio provides to applications running inside Istio.

> Many of the features the Kubernetes automatically provides applications are supplemented by Istio.

Author: Basically, that K8s gives a bunch of functionality for free. When you are on a VM, you are missing all of that (dns, health checks, etc). With Istio, you get those functionalities added to VMs *and* the standard istio value add (mtls, telemetry, etc).

This includes service discovery, DNS resolution, and health checks.
Additionally, standard Istio features such as automatic mutual TLS, rich telemetry, and expressive traffic management configuration are enabled.

The following diagram shows the architecture of a mesh with virtual machines:

{{< tabset category-name="network-mode" >}}

{{< tab name="Single-Network" category-value="single" >}}

> In this mesh, there is a single [network](/docs/ops/deployment/deployment-models/#network-models), meaning pods and virtual machines can directly reach each other.

Reviewer 2: ```suggestion
In this mesh, there is a single [network](/docs/ops/deployment/deployment-models/#network-models), implying that pods and virtual machines can communicate directly with each other.
```

> Control plane traffic, including XDS configuration and certificate signing, are sent through a Gateway in the cluster.

Reviewer 2: Is this an ingress/egress GW? Does this matter which one?

Author: It can be any gateway with istiod exposed. It doesn't even need to be an istio gateway technically. Our docs recommend a dedicated gateway.

> This ensures that the VMs have a stable address to connect to when they are bootstrapping. All other communication between pods and

Reviewer 2: ```suggestion
This ensures that the virtual machines have a stable address to connect to when they are bootstrapping. All other communication between pods and
```

> This ensures that the VMs have a stable address to connect to when they are bootstrapping. All other communication between pods and
> virtual machines are able to communicate directly.

Reviewer 2: ```suggestion
This ensures that the VMs have a stable address to connect to when they are bootstrapping. Pods and virtual machines can communicate directly with each other without requiring any intermediate Gateway.
```

{{< image width="75%"
    link="single-network.svg"
    alt="A service mesh with a single network and virtual machines"
    title="Single network"
    caption="A service mesh with a single network and virtual machines"
    >}}

{{< /tab >}}

{{< tab name="Multi-Network" category-value="multiple" >}}

> In this mesh, there is are multiple [networks](/docs/ops/deployment/deployment-models/#network-models), meaning pods and virtual machines are not able to directly reach each other.

Reviewer 2: ```suggestion
In this mesh, there are multiple [networks](/docs/ops/deployment/deployment-models/#network-models), implying that pods and virtual machines are not able to communicate directly with each other.
```

Control plane traffic, including XDS configuration and certificate signing, are sent through a Gateway in the cluster.
Similarly, all communication between pods and virtual machines goes through the gateway, which acts as a bridge between the two networks.

{{< image width="75%"
    link="multi-network.svg"
    alt="A service mesh with multiple networks and virtual machines"
    title="Multiple networks"
    caption="A service mesh with multiple networks and virtual machines"
    >}}

{{< /tab >}}

{{< /tabset >}}

## Service association

> Istio introduces two new types to represent virtual machine workloads:

Reviewer 2: ```suggestion
Istio provides two new mechanisms to represent virtual machine workloads:
```

> * [`WorkloadGroup`](/docs/reference/config/networking/workload-group/) represent a grouping of workloads that share common properties. This is comparable to a `Deployment` in Kubernetes

Reviewer 2: ```suggestion
* [`WorkloadGroup`](/docs/reference/config/networking/workload-group/) represent a logical group of virtual machine workloads that share common properties. This is similar to a `Deployment` in Kubernetes.
```

> * [`WorkloadEntry`](/docs/reference/config/networking/workload-entry/) represent a single instance of a workload. This is comparable to a `Pod` in Kubernetes.

Reviewer 2: ```suggestion
* [`WorkloadEntry`](/docs/reference/config/networking/workload-entry/) represent a single instance of a virtual machine workload. This is similar to a `Pod` in Kubernetes.
```

> Unlike their corresponding Kubernetes resources, creating these resources does not result in provisioning of any resources or running any applications.

Reviewer 2: ```suggestion
Creating these resources (WorkloadGroup and WorkloadEntry) does not result in provisioning of any resources or running any virtual machine workloads.
```

Rather, these resources just reference these workloads and inform Istio how to configure the mesh appropriately.

> When adding a virtual machine workload to the mesh, you will need to create a `WorkloadGroup`, that acts as template for each instance:

Reviewer 2: ```suggestion
When adding a virtual machine workload to the mesh, you will need to create a `WorkloadGroup`, that acts as template for each `WorkloadEntry` instance:
```

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: WorkloadGroup
metadata:
  name: product-vm
spec:
  metadata:
    labels:
      app: product
  template:
    serviceAccount: default
  probe:
    httpGet:
      port: 8080
{{< /text >}}

> Once the virtual machine has been [configured and added to the mesh](/docs/setup/install/virtual-machine/#configure-the-virtual-machine), a corresponding `WorkloadEntry` will be automatically created.

Reviewer 2: ```suggestion
Once the virtual machine has been [configured and added to the mesh](/docs/setup/install/virtual-machine/#configure-the-virtual-machine), a corresponding `WorkloadEntry` will be automatically created by the Istio control plane.
```

For example:

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: WorkloadEntry
metadata:
  annotations:
    istio.io/autoRegistrationGroup: product-vm
  labels:
    app: product
  name: product-vm-1.2.3.4
spec:
  address: 1.2.3.4
  labels:
    app: product
  serviceAccount: default
{{< /text >}}

> These `WorkloadEntry` describe a single instance of a workload, similar to a pod. When the workload is removed from the mesh, the `WorkloadEntry` will

Reviewer 2: ```suggestion
This `WorkloadEntry` resource describes a single instance of a workload, similar to a pod in Kubernetes. When the workload is removed from the mesh, the `WorkloadEntry` resource will
```

> be automatically removed, and its health status will be automatically updated based on the probe configured in the `WorkloadGroup`, if any.

Reviewer 2: ```suggestion
be automatically removed.  Additionally, if any probes are configured in the `WorkloadGroup` resource, the Istio control plane automatically updates the health status of associated `WorkloadEntry` instances. 
```

> In order for consumers to actually call the workload, generally it's desired to join a part of a `Service`. This allows
> clients to reach a stable hostname, like `product.default.svc.cluster.local`, rather than an IP address that may change.
> Additionally, it allows Istio configuration such as `DestinationRule`s and `VirtualService`s to apply.

Reviewer 2: ```suggestion
In order for consumers to reliably call your workload, it's recommended to declare a `Service` association. This allows clients to reach a stable hostname, like `product.default.svc.cluster.local`, rather than ephemeral IP addresses. This also enables you to use advanced routing capabilities in Istio via `DestinationRule` and `VirtualService` APIs.
```

> Joining a `Service` works the same as with `Pod`s; the `Service`'s selector will match over the `WorkloadEntry` labels.

Reviewer 2: ```suggestion
Any Kubernetes service transparently selects workloads across pods and virtual machines via the selector fields which are matched with pod and `WorkloadEntry` labels respectively.
```

> For example, if we had a `Service` named `product` that was composed of a `Pod` and a `WorkloadEntry`:

Reviewer 2: ```suggestion
For example, for a `Service` named `product` that was composed of a `Pod` and a `WorkloadEntry`:
```

{{< image width="30%"
    link="service-selector.svg"
    title="Service Selection"
    >}}

With this configuration, requests to `product` would be sent to both the pod and virtual machine workload.

## DNS

> One Kubernetes, pods will automatically be configured with a DNS resolver that enables `Service` names.
> This enables pods to easily communicate with one another by simple, stable hostnames.

Reviewer 2: ```suggestion
Kubernetes provides DNS resolution in pods for `Service` names  allowing pods to easily communicate with one another by stable hostnames.
```

> Istio provides similar functionality to virtual machine workloads through its [DNS Proxy](/docs/ops/configuration/traffic-management/dns-proxy/).
> This feature redirects all DNS queries from the workload through the Istio proxy, which maintains a mapping of hostnames to IP addresses.

Reviewer 2: ```suggestion
For virtual machine expansion Istio provides similar functionality via [DNS Proxy](/docs/ops/configuration/traffic-management/dns-proxy/).
This feature redirects all DNS queries from the virtual machine workload to the Istio proxy, which maintains a mapping of hostnames to IP addresses.
```

> As a result, workloads running on virtual machines will be able to call `Service`s just like pods.

Reviewer 2: ```suggestion
As a result, workloads running on virtual machines can transparently call `Service`s (similar to pods) without requiring any additional configuration.
```
